subsys/usb/device/class/cdc_acm.c

```diff
@@ -624,7 +624,8 @@ static int cdc_acm_irq_tx_ready(const struct device *dev)
 {
 	struct cdc_acm_dev_data_t * const dev_data = dev->data;
 
-	if (dev_data->tx_irq_ena && dev_data->tx_ready) {
+	if (dev_data->tx_irq_ena && dev_data->tx_ready && dev_data->configured &&
+	    !dev_data->suspended) {
```

Comment:
I already mentioned in another PR review that these changes cannot go in as they are. Also, did you read my comment about [issue](https://github.com/zephyrproject-rtos/zephyr/issues/76064#issuecomment-2237742867) you reported? What fixed it here for you is just the additional check for being configured, but not being configured should not be a reason to not trigger the TX interrupt as long as there is space in the ring buffer.

Reply:
I personally prefer data to be rejected when suspended / not configured, but that's not my decision.

My issue is: 
cdc_acm_irq_tx_ready()
and
cdc_acm_fifo_fill()
need to match. When cdc_acm_irq_tx_ready() is true cdc_acm_fifo_fill() needs to take at least one byte. Everything else violates the 
[UART API](https://docs.zephyrproject.org/2.7.5/reference/peripherals/uart.html#group__uart__interface_1ga5e126b5f19549eb7f5b785b98ebe7638).

> Check if UART TX buffer can accept at least one character for transmission (i.e. [uart_fifo_fill()](https://docs.zephyrproject.org/2.7.5/reference/peripherals/uart.html#group__uart__interface_1gafe42e4719eada7e25904bc9ebfe87791) will succeed and return non-zero).


So either cdc_acm_irq_tx_ready() changes as I proposed (matching cdc_acm_fifo_fill() not taking data) OR cdc_acm_fifo_fill() changes taking the data even though it can't be sent (as not configured / suspended). As I said, cdc_acm_fifo_fill() does it right in my opinion, but I am not willing to fight about it. Just let me fix the API issue whichever way you prefer.

As stated (https://github.com/zephyrproject-rtos/zephyr/pull/76171#discussion_r1687512260) I would use cdc_acm_irq_tx_ready() for the check in cdc_acm_fifo_fill() so this way it always matches even if the implementation of suspend gets changed later (or, even cleaner, make another (inline?) function called by both).